### PR TITLE
separate out 3rd-party headers

### PR DESCRIPTION
**Pull request by aws-nslick**
```
88569151 feat(build): handle 3p headers better

Don't put 3p dependencies under the same directory as our headers, put
them under a separate top-level 3rd-party directory. Substitute in the
accelerator (cuda or neuron), so that our code doesn't need to actually
care if it's neuron or nvidia at every inclusion, including <nccl/err.h>
will do the right thing either way. Put the tuner under its own
Makefile. Move uthash, too. Prefer <> to "" so that it's clear that
these are system headers, not code we control. Also helps silence
warnings stemming from these files.

Signed-off-by: Nicholas Sielicki <nslick@amazon.com>

43 files changed, 144 insertions(+), 131 deletions(-)
.gitignore                                         |  1 +
3rd-party/Makefile.am                              | 28 ++++++++++++++++++
.../nccl/cuda/include/nccl}/err.h                  |  0 
.../nccl/cuda/include/nccl}/net.h                  |  0 
.../nccl/cuda/include/nccl}/net_device.h           |  0 
.../nccl/cuda/include/nccl}/net_v2.h               |  0 
.../nccl/cuda/include/nccl}/net_v3.h               |  0 
.../nccl/cuda/include/nccl}/net_v4.h               |  0 
.../nccl/cuda/include/nccl}/net_v5.h               |  0 
.../nccl/cuda/include/nccl}/net_v6.h               |  0 
.../nccl/cuda/include/nccl}/net_v7.h               |  0 
.../nccl/cuda/include/nccl}/net_v8.h               |  0 
.../nccl/cuda/include/nccl}/tuner.h                |  4 +--
.../nccl/cuda/include/nccl}/tuner_v1.h             |  0 
.../nccl/cuda/include/nccl}/tuner_v2.h             |  0 
.../nccl/cuda/include/nccl}/types.h                |  0 
3rd-party/nccl/neuron/include/nccl/err.h           |  1 +
.../nccl/neuron/include/nccl}/error.h              |  0 
.../nccl/neuron/include/nccl}/net.h                |  0 
.../nccl/neuron/include/nccl}/net_v4.h             |  0 
.../nccl/neuron/include/nccl}/net_v5.h             |  0 
.../uthash/include/uthash}/uthash.h                |  0 
.../uthash/include/uthash}/utlist.h                |  0 
Makefile.am                                        |  2 +-
configure.ac                                       |  4 +++
include/Makefile.am                                | 22 --------------
include/nccl-headers/error.h                       | 16 ----------
include/nccl-headers/net.h                         | 16 ----------
include/nccl_ofi.h                                 | 14 +++++----
include/nccl_ofi_api.h                             |  4 +--
include/nccl_ofi_log.h                             |  2 +-
include/nccl_ofi_rdma.h                            | 12 ++++----
include/nccl_ofi_sendrecv.h                        |  4 +--
include/nccl_ofi_tuner.h                           |  2 +-
src/Makefile.am                                    | 34 +++++++---------------
src/nccl_ofi_ep_addr_list.c                        |  9 +++---
src/nccl_ofi_sendrecv.c                            |  1 +
src/tuner/Makefile.am                              | 32 ++++++++++++++++++++
src/tuner/nccl_ofi_tuner.c                         |  8 ++---
tests/functional/Makefile.am                       |  5 +++-
tests/functional/test-common.h                     | 19 +++++++-----
tests/unit/Makefile.am                             | 26 +++++++++++------
tests/unit/scheduler.c                             |  9 +++---
```

```
6df3fb59 feat(build): use -isystem instead of -I

When constructing include directories, use -isystem instead of -I. This
accomplishes largely the same thing, but informs the compiler that we
don't control these headers, which is useful for filtering warnings.

Signed-off-by: Nicholas Sielicki <nslick@amazon.com>
```

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by aws-nslick** [comment]
Moved to draft, was hoping this could avoid being based atop other work but it's not possible.

File: tests/unit/Makefile.am [hunk 1]
```diff
@@ -5,31 +5,39 @@
...
 	scheduler \
 	idpool \
 	ep_addr_list \
 	mr
 
-TESTS = $(noinst_PROGRAMS)
-
-show_tuner_decisions_SOURCES = show_tuner_decisions.c
-show_tuner_decisions_LDADD = $(top_builddir)/src/libinternal_tuner_plugin.la
+if !ENABLE_NEURON
```
Comment (aws-nslick) [inline tests/unit/Makefile.am:23]: incorrect?

File: 3rd-party/nccl/neuron/include/nccl/err.h [hunk 2]
```diff
@@ -0,0 +1 @@
+error.h
```
Comment (bwbarrett) [inline 3rd-party/nccl/neuron/include/nccl/err.h:1]: ?

Comment (aws-nslick) [reply 3rd-party/nccl/neuron/include/nccl/err.h:1]: nccl does not ship error.h, they ship err.h:  https://github.com/NVIDIA/nccl/blob/master/ext-net/example/nccl/err.h

neuron does not ship err.h, they ship error.h: https://github.com/aws/aws-ofi-nccl/tree/master/include/nccl-headers/neuron

this is how github displays symlinks.

File: tests/unit/Makefile.am (hunk 1 above)
Comment (aws-nslick) [reply tests/unit/Makefile.am:23]: no, neuron does not contain a tuner.h interface file, tuner cannot be built for neuron until they opt in and vend a copy.

File: 3rd-party/nccl/neuron/include/nccl/err.h (hunk 2 above)
Comment (bwbarrett) [reply 3rd-party/nccl/neuron/include/nccl/err.h:1]: Ah, that's... lame of github.

**Review by bwbarrett — APPROVED** [review]

**Review by bwbarrett — APPROVED** [review]

**Pull request merged**